### ISSUE TITLE
Adding TrackedReal is ambiguous in HMC

### ISSUE THREAD

**Issue by MrVPlusOne** [probcomp/Gen.jl#420]
Hi, when I was trying to use HMC in Gen, I got the following MethodError related to addition between TrackedReals. Since I'm not sure if this is a general issue of ReverseDiff or a specific problem in Gen's implementation, I'm posting the issue here. Have you seen similar issues before? Is there a quick way to fix this? Many thanks! 

```
ERROR: LoadError: MethodError: (::ReverseDiff.ForwardOptimize{typeof(+)})(::ReverseDiff.TrackedReal{Float64, Float64, Nothing}, ::ReverseDiff.TrackedReal{ReverseDiff.TrackedReal{Float64, Float64, Nothing}, ReverseDiff.TrackedReal{Float64, Float64, Nothing}, Nothing}) is ambiguous. Candidates:
  (self::ReverseDiff.ForwardOptimize{F})(t::ReverseDiff.TrackedReal{V, D, O} where O, x::Real) where {F, V, D} in ReverseDiff at /Users/jiayiwei/.julia/packages/ReverseDiff/60noS/src/macros.jl:121
  (self::ReverseDiff.ForwardOptimize{F})(x::Real, t::ReverseDiff.TrackedReal{V, D, O} where O) where {F, V, D} in ReverseDiff at /Users/jiayiwei/.julia/packages/ReverseDiff/60noS/src/macros.jl:109
Possible fix, define
  (::ReverseDiff.ForwardOptimize{F})(::ReverseDiff.TrackedReal{V, D, O} where O, ::ReverseDiff.TrackedReal{V, D, O} where O) where {F, V, D, V, D}
Stacktrace:
 [1] +(a::ReverseDiff.TrackedReal{Float64, Float64, Nothing}, b::ReverseDiff.TrackedReal{ReverseDiff.TrackedReal{Float64, Float64, Nothing}, ReverseDiff.TrackedReal{Float64, Float64, Nothing}, Nothing})
   @ ReverseDiff ~/.julia/packages/ReverseDiff/60noS/src/derivatives/scalars.jl:9
```

**Comment by ztangent** [comment]
Can you post the code of the `@gen` function that you ran HMC on? This may be related to #314, but it's hard to be sure until I see more of the function!

**Comment by MrVPlusOne** [comment]
Sure, the `@gen` function samples some parameterized dynamics and uses it to forward simulate a dynamical system.
```julia
@gen function data_process(
    num_params::Int, s1, actions, times, δ_state_f
)
    params = [{(:param, i)} ~ normal(0.0, 1.0) for i in 1:num_params]
    dynamics = to_dynamics(params, compiled)
    δ_scale = {:δ_scale} ~ inv_gamma(3.0, 1.0)
    state = s1
    for t in 1:length(times)-1
        ŝ = dynamics(state, actions[t], times[t], times[t+1]-times[t])
        state = {(:state, t+1)} ~ δ_state_f(ŝ, δ_scale)
    end
end
```

Then, I generate a trace with a choicemap containing the observed states at every time step and perform inference using hmc.
```julia
constraints = choicemap(((:state, t) => traj[t].rocket for t in 2:length(traj))...)
trace = generate(data_process, args, constraints)[1]
...
for i in 1:num_steps
    trace, accepted = hmc(trace, sel; L=10, eps)
    ...
end
```

The error is then thrown at the observation line `state = {(:state, t+1)} ~ δ_state_f(ŝ, δ_scale)`, where `δ_state_f` is a custom distribution defined as
```julia
struct RocketStateNoise <: Gen.Distribution{RocketState} end
Gen.logpdf(::RocketStateNoise, v::RocketState, μ::RocketState, σ) = 
    sum(Gen.logpdf.(Ref(normal), rocket_to_vector(v-μ), 0.0, σ))
Gen.random(::RocketStateNoise, μ::RocketState, σ)::RocketState{ℝ} = 
    vector_to_rocket(SVector{6}(normal(m, σ) for m in rocket_to_vector(μ)))
```
Here `vector_to_rocket` and `rocket_to_vector` convert a `RocketState` from and to a `StaticVector`.

**Comment by ztangent** [comment]
Okay, my take on this is that it's not related to #314. I can't say more without knowing more about how `RocketStateNoise` is implemented, given that you're sampling from that distribution. Do you have `Gen.logpdf_grad` defined for `RocketStateNoise`?

(Also, I think this is probably unrelated to your issue, but you might be able to use `broadcasted_normal` in place of `normal` to handle multi-variate normal distributions with diagonal covariance matrices.)

**Comment by MrVPlusOne** [comment]
The last code block contains all the code I have for `RocketStateNoise`, and I have not defined a`Gen.logpdf_grad` function for it. (I thought it would be automatically handled by the AD framework, is that not the case?) Also, given that I constrained all the state observations using the choicemap, I believe only `logpdf` (and no sampling code) involving `RocketStateNoise ` is used during inference. 

Do you have any suggestions on what I should try next to fix this? Should I open an issue on ReverseDiff?

> (Also, I think this is probably u #nrelated to your issue, but you might be able to use broadcasted_normal in place of normal to handle multi-variate normal distributions with diagonal covariance matrices.)

I was looking for that function but didn't find it. Thanks for the info!

**Comment by ztangent** [comment]
Hmm, I'm not 100% sure, but I do think you need to implement `logpdf_grad` to support gradient-based inference like HMC with custom distributions. It wouldn't be hard to implement, since I think you can just forward the call to `logpdf_grad` for `broadcasted_normal`. But I haven't used gradient-based inference very much in Gen, so @marcoct or @alex-lew might know better.

**Comment by MrVPlusOne** [comment]
> but I do think you need to implement logpdf_grad to support gradient-based inference like HMC with custom distributions.

Hmm...I think you are right. The weird thing is that after I replaced `params = [{(:param, i)} ~ normal(0.0, 1.0) for i in 1:num_params]` with `params = {:params} ~ broadcasted_normal(zeros(n_params), ones(n_params))`, it now starts to complain `ERROR: LoadError: MethodError: no method matching logpdf_grad(::ProbPRL.RocketStateNoise, ::RocketState{Float64, Float64, Float64, Float64}, ::RocketState{Float64, Float64, Float64, Float64}, ::Float64)`!

With `broadcasted_normal`, I then tried to redefine `rocket_state_noise` using `@dist` as
```julia
@dist rocket_noise_dist(s::RocketState, σ) =
    (rocket_to_vector(s) + broadcasted_normal(zeros(6), ones(6) * σ)) |> vector_to_rocket
```
But unfortunately, this did not work
```
┌ Error: Failed to revise /Users/jiayiwei/.julia/dev/ProbPRL.jl/src/Synthesis.jl
│   exception =
│    MethodError: no method matching +(::Gen.TransformedArg{Any}, ::Gen.DistWithArgs{Array{Float64, N} where N})
│    Closest candidates are:
│      +(::Any, ::Any, ::Any, ::Any...) at operators.jl:560
│      +(::Union{InitialValues.NonspecificInitialValue, InitialValues.SpecificInitialValue{typeof(+)}}, ::Any) at /Users/jiayiwei/.julia/packages/InitialValues/EPz1F/src/InitialValues.jl:153
│      +(::Diffed{var"#92#T", NoChange}, ::Any) where var"#92#T" at /Users/jiayiwei/.julia/packages/Gen/unzLD/src/diff.jl:289
│      ...
│    Stacktrace:
│     [1] dist_call(f::Function, args::Tuple{Gen.TransformedArg{Any}, Gen.DistWithArgs{Array{Float64, N} where N}}, __module__::Module)
│       @ Gen ~/.julia/packages/Gen/unzLD/src/modeling_library/dist_dsl/dist_dsl.jl:118
│     [2] top-level scope
│       @ ~/.julia/dev/ProbPRL.jl/src/Synthesis.jl:137
```

It looks like `rocket_to_vector(s)` is inferred to have the type `Gen.TransformedArg{Any}`. Does the `@dist` DSL have additional restrictions on which functions can be used inside the body or only handle a specific subset of Juila data types?

**Comment by ztangent** [comment]
Maybe try:
```julia
@dist rocket_noise_dist(s::RocketState, σ) =
    broadcasted_normal(rocket_to_vector(s), ones(6) * σ) |> vector_to_rocket
```
and see if that works? I think there are some limitations of the `@dist` DSL you're running up against, though I'm not sure what off the top of me head. If that still doesn't work, maybe it's worth implementing `logpdf_grad` for `RocketNoiseDist` manually.

**Comment by MrVPlusOne** [comment]
@ztangent  Thank you so much for the help! I've managed to make it work by avoiding custom distributions altogether using `broadcasted_norm` (and refactor my state observations to be vectors), as follows:
```julia
ŝ = dynamics(state, actions[t], times[t], Δt) |> rocket_to_vector
state = vector_to_rocket( {(:state, t+1)} ~ broadcasted_normal(ŝ, abs(δ_scale) * Δt))
```
However, I did encounter two additional problems and (I think) managed to fix them. One of them I believe is a bug, the other one I'm not sure.

The first problem I encountered is that the `logpdf_grad` implementation of  `broadcasted_norm` seems to only handle the cases where *all* arguments are scalars (note that it returns 3 scalars as gradients):
```julia
function logpdf_grad(::BroadcastedNormal,
                     x::Union{AbstractArray{<:Real}, Real},
                     mu::Union{AbstractArray{<:Real}, Real},
                     std::Union{AbstractArray{<:Real}, Real})
    assert_has_shape(x, broadcast_shapes_or_crash(mu, std);
                     msg="Shape of `x` does not agree with the sample space")
    precision = 1.0 ./ (std .* std)
    diff = mu .- x
    deriv_x = sum(diff .* precision)
    deriv_mu = sum(-deriv_x)
    deriv_std = sum(-1.0 ./ std .+ (diff .* diff) ./ (std .* std .* std))
    (deriv_x, deriv_mu, deriv_std)
end
```
This would give incorrect gradient for vector inputs, and I believe the correct implementation should be something like this:
```julia
function Gen.logpdf_grad(
    ::Gen.BroadcastedNormal,
    x::Union{AbstractArray{<:Real}, Real},
    mu::Union{AbstractArray{<:Real}, Real},
    std::Union{AbstractArray{<:Real}, Real}
)
    Gen.assert_has_shape(x, Gen.broadcast_shapes_or_crash(mu, std);
    msg="Shape of `x` does not agree with the sample space")
    precision = 1.0 ./ (std .* std)
    diff = mu .- x
    deriv_x = diff .* precision
    deriv_mu = -deriv_x
    deriv_std = -1.0 ./ std .+ (diff .* diff) ./ (std .* std .* std)
    ((x isa Real) ? sum(deriv_x) : deriv_x, 
    (mu isa Real) ? sum(deriv_mu) : deriv_mu, 
    (std isa Real) ? sum(deriv_std) : deriv_std)
end
```
After fixing the above issue using method overriding, I encountered the following MethodError:
```
ERROR: LoadError: MethodError: no method matching increment_deriv!(::Float64, ::Float64)
Closest candidates are:
  increment_deriv!(::ReverseDiff.TrackedArray, ::Real, ::Any) at /Users/jiayiwei/.julia/packages/ReverseDiff/E4Tzn/src/derivatives/propagation.jl:34
  increment_deriv!(::AbstractArray, ::Real, ::Any) at /Users/jiayiwei/.julia/packages/ReverseDiff/E4Tzn/src/derivatives/propagation.jl:36
  increment_deriv!(::AbstractArray, ::Any) at /Users/jiayiwei/.julia/packages/ReverseDiff/E4Tzn/src/derivatives/propagation.jl:38
  ...
Stacktrace:
  [1] increment_deriv!
    @ ~/.julia/packages/ReverseDiff/E4Tzn/src/derivatives/propagation.jl:35 [inlined]
  [2] increment_deriv!(t::StaticArrays.SVector{6, Real}, x::Vector{Float64})
    @ ReverseDiff ~/.julia/packages/ReverseDiff/E4Tzn/src/derivatives/propagation.jl:40
  [3] increment_deriv!(arg::StaticArrays.SVector{6, Real}, deriv::Vector{Float64})
    @ Gen ~/.julia/packages/Gen/unzLD/src/backprop.jl:16
  [4] special_reverse_exec!(instruction::ReverseDiff.SpecialInstruction{Gen.BroadcastedNormal, Tuple{Vector{Float64}, StaticArrays.SVector{6, Real}, ReverseDiff.TrackedReal{Float64, Float64, Nothing}}, ReverseDiff.TrackedReal{Float64, Float64, Nothing}, Nothing})
    @ Gen ~/.julia/packages/Gen/unzLD/src/dynamic/backprop.jl:24
  [5] reverse_exec!(instruction::ReverseDiff.SpecialInstruction{Gen.BroadcastedNormal, Tuple{Vector{Float64}, StaticArrays.SVector{6, Real}, ReverseDiff.TrackedReal{Float64, Float64, Nothing}}, ReverseDiff.TrackedReal{Float64, Float64, Nothing}, Nothing})
    @ ReverseDiff ~/.julia/packages/ReverseDiff/E4Tzn/src/tape.jl:93
  [6] reverse_pass!(tape::Vector{ReverseDiff.AbstractInstruction})
    @ ReverseDiff ~/.julia/packages/ReverseDiff/E4Tzn/src/tape.jl:87
  [7] reverse_pass!
    @ ~/.julia/packages/Gen/unzLD/src/backprop.jl:12 [inlined]
  [8] choice_gradients(trace::Gen.DynamicDSLTrace{DynamicDSLFunction{Any}}, selection::DynamicSelection, retval_grad::Nothing)
    @ Gen ~/.julia/packages/Gen/unzLD/src/dynamic/backprop.jl:382
  [9] hmc(trace::Gen.DynamicDSLTrace{DynamicDSLFunction{Any}}, selection::DynamicSelection; L::Int64, eps::Float64, check::Bool, observations::EmptyChoiceMap)
```
Looking at stack trace [2], it looks like it's caused by Gen calling `increment_deriv!(t::StaticArrays.SVector{6, Real}, x::Vector{Float64})`, which causes it to call `increment_deriv!` on every element of the vector. But there is no implementation for `increment_deriv!(::Real, ::Real)` (only has an implementation for the `(::TrackedReal, ::Real)` case). So I added this definition
```julia
import ReverseDiff
ReverseDiff.increment_deriv!(::Float64, ::Float64) = nothing
```
and now hmc seems to be working properly (from the look of the inference results), but I'm not sure if my fix is correct and whether this is supposed to happen. I tried to reproduce this bug using a simpler example involving `broadcasted_norm` but have not succeeded so far.

**Comment by ztangent** [comment]
Glad that HMC seems to be working!

Your fix to `logpdf_grad` seems correct, though I'll defer to @bzinberg who I believe implemented `broadcasted_normal`.

As for the `increment_deriv!` issue, you might want to try updating your version of `ReverseDiff` -- that particular issue seems like it was an upstream bug that has since been resolved (see #415).

**Comment by MrVPlusOne** [comment]
> As for the increment_deriv! issue, you might want to try updating your version of ReverseDiff -- that particular issue seems like it was an upstream bug that has since been resolved (see #415).

Thanks for the info! I tried to update ReverseDiff to 1.9.0, but unfortunately, the problem still persists for me—it seems to have something to do with `StaticArrays` and `broadcasted_norm`.  If I collect the `StaticArray` before feeding them to `broadcasted_norm`, the problem went away in 1.9.0, and the performance is better than using `StaticArray` with overriding `increment_deriv!(::Float64, ::Float64) = nothing`.

> Your fix to logpdf_grad seems correct, though I'll defer to @bzinberg who I believe implemented broadcasted_normal.

Actually, I realized that we will need a more general fix to handle all possible broadcasting scenarios. My new fix now looks like this
```julia
function Gen.logpdf_grad(
    ::Gen.BroadcastedNormal,
    x::Union{AbstractArray{<:Real}, Real},
    mu::Union{AbstractArray{<:Real}, Real},
    std::Union{AbstractArray{<:Real}, Real}
)
    Gen.assert_has_shape(x, Gen.broadcast_shapes_or_crash(mu, std);
    msg="Shape of `x` does not agree with the sample space")
    precision = 1.0 ./ (std .* std)
    diff = mu .- x
    deriv_x = diff .* precision
    deriv_mu = -deriv_x
    deriv_std = -1.0 ./ std .+ (diff .* diff) ./ (std .* std .* std)
    (unbroadcast_for_arg(x, deriv_x), 
    unbroadcast_for_arg(mu, deriv_mu), 
    unbroadcast_for_arg(std, deriv_std))
end

unbroadcast_for_arg(::Real, grad) = sum(grad)
function unbroadcast_for_arg(
    arg::AbstractArray{<:Real, N}, grad::AbstractArray{T}
)::AbstractArray{T, N} where {N,T}
    size(arg) == size(grad) ? grad : unbroadcast_grad(size(arg), grad)
end

function unbroadcast_grad(
    old_shape::NTuple{l_old, Int}, grad::AbstractArray{T, l_new}
) where {T, l_old, l_new}
    @assert l_new >= l_old  
    new_shape = size(grad)
    dims=filter(i -> i > l_old || old_shape[i] == 1 && new_shape[i] > 1, 1:l_new)
    dropdims(sum(grad; dims); dims=tuple((l_old+1:l_new)...))::AbstractArray{T, l_old}
end
```
Hope this is helpful!

**Comment by bzinberg** [comment]
> the `logpdf_grad` implementation of `broadcasted_norm` seems to only handle the cases where _all_ arguments are scalars (note that it returns 3 scalars as gradients)

Good catch, thanks @MrVPlusOne!  Indeed, in https://github.com/probcomp/Gen.jl/commit/005f97c7176ba19cd6f6b88d00cfa2cb32f25c92#, I blindly inserted a `sum` into both `logpdf` and `logpdf_grad`... whoops.  And indeed, what `logpdf_grad` should instead do is sum over the dims that are "missing" from each arg, compared to the fully broadcasted shape, as you do above.

@MrVPlusOne, if you're up for sending a PR with this fix, I'd be happy to review it!
